db/blob/blob_source.cc

```diff
@@ -0,0 +1,212 @@
+                                    CachableEntry<Slice>* cached_blob,
+                                    PinnableSlice* blob) const {
+  assert(blob);
+  assert(!cache_key.empty());
+
+  const Cache::Priority priority = Cache::Priority::LOW;
+
+  Status s;
+
+  if (blob_cache != nullptr) {
+    Cache::Handle* cache_handle = nullptr;
+    s = InsertEntryIntoCache(blob_cache, cache_key, blob, blob->size(),
+                             &cache_handle, priority);
+    if (s.ok()) {
+      assert(cache_handle != nullptr);
+      cached_blob->SetCachedValue(blob, blob_cache, cache_handle);
+    }
+  }
+
+  return s;
+}
+
+Status BlobSource::GetBlob(const ReadOptions& read_options,
+                           const Slice& user_key, uint64_t file_number,
+                           uint64_t offset, uint64_t file_size,
+                           uint64_t value_size,
+                           CompressionType compression_type,
+                           FilePrefetchBuffer* prefetch_buffer,
+                           PinnableSlice* value, uint64_t* bytes_read) {
+  assert(value);
+
+  const uint64_t key_size = user_key.size();
+
+  if (!IsValidBlobOffset(offset, key_size, value_size, file_size)) {
+    return Status::Corruption("Invalid blob offset");
+  }
+
+  const CacheKey cache_key = GetCacheKey(file_number, file_size, offset);
+
+  CachableEntry<Slice> blob_entry;
+
+  Status s;
+  if (blob_cache_) {
+    s = MaybeReadBlobFromCache(prefetch_buffer, read_options, cache_key, offset,
+                               &blob_entry);
+
+    if (s.ok() && blob_entry.GetValue() != nullptr) {
+      assert(blob_entry.GetValue()->size() == value_size);
+      if (bytes_read) {
+        *bytes_read = value_size;
+      }
+      value->PinSelf(*blob_entry.GetValue());
+      return s;
+    }
+  }
+
+  assert(blob_entry.IsEmpty());
+
+  const bool no_io = read_options.read_tier == kBlockCacheTier;
+  if (no_io) {
+    return Status::Incomplete("Cannot read blob(s): no disk I/O allowed");
+  }
+
+  // Can't find the blob from the cache. Since I/O is allowed, read from the
+  // file.
+  {
+    CacheHandleGuard<BlobFileReader> blob_file_reader;
+    s = blob_file_cache_->GetBlobFileReader(file_number, &blob_file_reader);
+    if (!s.ok()) {
+      return s;
+    }
+
+    assert(blob_file_reader.GetValue());
+
+    if (compression_type != blob_file_reader.GetValue()->GetCompressionType()) {
+      return Status::Corruption("Compression type mismatch when reading blob");
+    }
+
+    s = blob_file_reader.GetValue()->GetBlob(
+        read_options, user_key, offset, value_size, compression_type,
+        prefetch_buffer, value, bytes_read);
+    if (!s.ok()) {
+      return s;
+    }
+  }
+
+  if (blob_cache_ && read_options.fill_cache) {
+    // If filling cache is allowed and a cache is configured, try to put the
+    // blob to the cache.
+    Slice key = cache_key.AsSlice();
+    s = PutBlobIntoCache(key, blob_cache_.get(), &blob_entry, value);
```

Comment:
Just to recap, there are some important points about objects put in the cache:
1) They have to be self-contained, i.e. own their contents.
2) They have to be heap-allocated.
3) The `Cache` has to be able to take unique ownership of them.

`PinnableSlice` can check the first box if it points to its own buffer; however, the application provided `PinnableSlice` object pointed to by `value` potentially/probably does not satisfy the second criterion, and definitely doesn't satisfy the third one. One way to fix this would be to put copy the blob read from the file into a *separate* heap-allocated `PinnableSlice`, and insert *that* into the cache. (Regarding the first point, note that a plain `Slice` is little more than a non-owning pointer and is thus not suitable for the purposes of `Cache`.)

Reply:
Did you mean to copy a `PinnableSlice` and put it into cache? Since `PinnableSlice` is self-contained (It can PinSelf).